src/main/java/cloud/eppo/ConfigurationWrangler.java

```diff
@@ -0,0 +1,241 @@
+package cloud.eppo;
+
+import cloud.eppo.api.Configuration;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ConfigurationWrangler {
+  private static final Logger log = LoggerFactory.getLogger(ConfigurationWrangler.class);
+
+  private final ConfigurationRequestor requestor;
+  private final IConfigurationStore configurationStore;
+
+  // This reference is **the** reference to the current configuration.
+  private volatile Configuration config;
+
+  ConfigurationWrangler(
+      ConfigurationRequestor requestor,
+      IConfigurationStore persistentSource,
+      Configuration initialConfiguration) {
+    this.requestor = requestor;
+    this.configurationStore = persistentSource;
+
+    setConfig(initialConfiguration != null ? initialConfiguration : Configuration.emptyConfig());
+  }
+
+  private final ReentrantLock lock = new ReentrantLock();
+
+  private void setConfig(Configuration configuration) {
+    lock.lock();
+    try {
+      config = configuration;
+
+      // Fire-and-forget save call to the config store. It may or may not block (non-blocking is
+      // best)
+      configurationStore.saveConfiguration(config);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  /**
+   * Loads the configuration from the `ConfigurationRequestor` and saves to "persistent" storage
+   * using `IConfigurationStore`
+   *
+   * <p>This method attempts to load from the `IConfigurationStore` if it was unable to load from
```

Comment:
~Our old implementation had the two race, and I think that may still be the approach we want to take~ Ah I see, this is a fallback.

I think we want to take a page from the JavaScript client SDK here (see [it's docs](https://docs.geteppo.com/sdks/client-sdks/javascript/#configuration-cache))

Basically, configuration stors have an `isExpired()` method. If we have a non-expired store, we just use that (no fetch necessary). If it is expired, we fetch. 

A few configuration options dictate behavior. If `useExpired` is true, then it will initialize right away. It will still fetch, and update based on `updateOnFetch` value.

This is all pretty complicated, so if it's too much to mirror the JavaScript client we don't have to. Also the iOS SDK doesn't either. However, consistency there would be good so while messing around this could be the right time to do it.

Reply:
Agreed this is a good time, but let's separate that from this PR for now and I'll send a follow up right behind it.